index.js

```diff
@@ -135,48 +140,67 @@ Filter.prototype.build = function() {
   instrumentation.stop();
   var plugin = this;
 
+  // used with options.async = true to allow 'create' and 'change' operations to complete async
+  var pendingWork = [];
+
   return new Promise(function(resolve) {
     resolve(heimdall.node('applyPatches', ApplyPatchesSchema, function(instrumentation) {
       var prevTime = process.hrtime();
-
-      var result = mapSeries(patches, function(patch) {
-        var operation = patch[0];
-        var relativePath = patch[1];
-        var entry = patch[2];
-        var outputPath = destDir + '/' + (plugin.getDestFilePath(relativePath, entry) || relativePath);
-        var outputFilePath = outputPath;
-
-        plugin._logger.debug('[operation:%s] %s', operation, relativePath);
-
-        switch (operation) {
-          case 'mkdir': {
-            instrumentation.mkdir++;
-            return fs.mkdirSync(outputPath);
-          } case 'rmdir': {
-            instrumentation.rmdir++;
-            return fs.rmdirSync(outputPath);
-          } case 'unlink': {
-            instrumentation.unlink++;
-            return fs.unlinkSync(outputPath);
-          } case 'change': {
-            instrumentation.change++;
-            return plugin._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, true, instrumentation);
-          } case 'create': {
-            instrumentation.create++;
-            return plugin._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, false, instrumentation);
-          } default: {
-            instrumentation.other++;
+      return new Promise(function(resolve) {
+        var result = mapSeries(patches, function(patch) {
+          var operation = patch[0];
+          var relativePath = patch[1];
+          var entry = patch[2];
+          var outputPath = destDir + '/' + (plugin.getDestFilePath(relativePath, entry) || relativePath);
+          var outputFilePath = outputPath;
+
+          plugin._logger.debug('[operation:%s] %s', operation, relativePath);
+
+          switch (operation) {
+            case 'mkdir': {
+              instrumentation.mkdir++;
+              return fs.mkdirSync(outputPath);
+            } case 'rmdir': {
+              instrumentation.rmdir++;
+              return fs.rmdirSync(outputPath);
+            } case 'unlink': {
+              instrumentation.unlink++;
+              return fs.unlinkSync(outputPath);
+            } case 'change': {
+              instrumentation.change++;
```

Comment:
We should move the instrumentation to where the work happens. That way, instrumentation will be accurate even if we early exit.